cmd/k8s-nameserver/main.go

```diff
@@ -0,0 +1,230 @@
+	toWatch := filepath.Join(defaultDNSConfigDir, kubeletMountedConfigLn)
+	go func() {
+		logger("starting file watch for %s", defaultDNSConfigDir)
+		if err != nil {
+			log.Fatalf("error starting a new configfile watcher: %v", err)
+		}
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					logger("watcher finished")
+					cancel()
+					return
+				}
+				if event.Name == toWatch {
+					msg := fmt.Sprintf("config update received: %s", event)
+					logger(msg)
+					c <- msg
+				}
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					logger("errors watcher finished: %v", err)
+					cancel()
+					return
+				}
+				if err != nil {
+					logger("error watching directory: %w", err)
+					cancel()
+					return
+				}
+			}
+		}
+	}()
+	if err = watcher.Add(defaultDNSConfigDir); err != nil {
+		log.Fatalf("failed setting up file watch for DNS config: %v", err)
+	}
+
+	ns := &nameserver{
+		configReader:  configReader,
+		configWatcher: c,
+		logger:        logger,
+		res:           res,
+	}
+
+	if err := ns.run(ctx, cancel); err != nil {
+		log.Fatalf("error running nameserver: %v", err)
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", udpEndpoint)
+	if err != nil {
+		log.Fatalf("error resolving UDP address: %v", err)
+	}
+	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		log.Fatalf("error opening udp connection: %v", err)
+	}
+	go func() {
+		<-ctx.Done()
+		conn.Close()
+	}()
+
+	logger("k8s-nameserver listening on: %v", addr)
+
+	for {
+		payloadBuff := make([]byte, 10000)
+		metadataBuff := make([]byte, 512)
+		_, _, _, addr, err := conn.ReadMsgUDP(payloadBuff, metadataBuff)
+		if err != nil {
+			logger(fmt.Sprintf("error reading UDP message: %v", err))
+			continue
+		}
+		dnsAnswer, err := ns.query(ctx, payloadBuff, addr.AddrPort())
+		if err != nil {
+			logger(fmt.Sprintf("error querying internal resolver: %v", err))
+			// reply with the dnsAnswer anyway
+		}
+		conn.WriteToUDP(dnsAnswer, addr)
+	}
+}
+
+type nameserver struct {
+	configReader  configReaderFunc
+	configWatcher <-chan string
+	res           *resolver.Resolver
+	logger        logger.Logf
+}
+
+type configReaderFunc func() ([]byte, error)
+
+// run ensures that resolver configuration is up to date with regards to its
+// source. will update config once before returning and keep monitoring it in a
+// thread.
+func (n *nameserver) run(ctx context.Context, cancelF context.CancelFunc) error {
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				n.logger("nameserver exiting")
+				return
+			case <-n.configWatcher:
+				n.logger("attempting to update resolver config...")
+				if err := n.updateResolverConfig(); err != nil {
+					n.logger("error updating resolver config: %w", err)
+					cancelF()
+				}
+				n.logger("successfully updated resolver config")
+			}
+		}
+	}()
+	if err := n.updateResolverConfig(); err != nil {
+		return fmt.Errorf("error updating resolver config: %w", err)
+	}
```

Comment:
I had to check if it's safe to concurrently call `updateResolverConfig` (it is), but to avoid other people wondering this, perhaps we could move this call to the beginning of the function, before the watcher goroutine is started?